cloud/scope/vpc_cluster.go

```diff
@@ -686,3 +742,326 @@ func (s *VPCClusterScope) buildCOSObjectHRef() (*string, error) {
+			subnetMap = s.NetworkStatus().ControlPlaneSubnets
+		} else if !isControlPlane && s.NetworkStatus().WorkerSubnets != nil {
+			subnetMap = s.NetworkStatus().WorkerSubnets
+		}
+		// Based on Network Status, setup either the name or ID for lookup of the subnet's current status.
+		if subnet.Name != nil {
+			if _, ok := subnetMap[*subnet.Name]; ok {
+				subnetName = subnet.Name
+			}
+		} else if subnet.ID != nil {
+			for _, statusSubnet := range subnetMap {
+				if statusSubnet.ID == *subnet.ID {
+					subnetID = subnet.ID
+				}
+			}
+		}
+
+		// Perform current status lookup of subnet, using ID or name if one was found in Network Status.
+		if subnetID != nil {
+			options := &vpcv1.GetSubnetOptions{
+				ID: subnetID,
+			}
+			subnetDetails, _, err := s.VPCClient.GetSubnet(options)
+			if err != nil {
+				return false, fmt.Errorf("error retrieving existing subnet by id %s: %w", *subnetID, err)
+			} else if subnetDetails == nil {
+				return false, fmt.Errorf("error failed to find existing subnet by id %s", *subnetID)
+			}
+			return s.updateSubnetStatus(subnetDetails, isControlPlane)
+		} else if subnetName != nil {
+			subnetDetails, err := s.VPCClient.GetVPCSubnetByName(*subnetName)
+			if err != nil {
+				return false, fmt.Errorf("error retrieving existing subnet by name %s: %w", *subnetName, err)
+			} else if subnetDetails == nil {
+				return false, fmt.Errorf("error failed to find existing subnet by name: %s", *subnetName)
+			}
+			return s.updateSubnetStatus(subnetDetails, isControlPlane)
+		}
+	}
+
+	// Otherwise, if these is an ID or name, attempt to lookup the subnet and update status as necessary.
+	if subnet.ID != nil {
+		options := &vpcv1.GetSubnetOptions{
+			ID: subnet.ID,
+		}
+		subnetDetails, _, err := s.VPCClient.GetSubnet(options)
+		if err != nil {
+			return false, fmt.Errorf("error retrieving subnet by id %s: %w", *subnet.ID, err)
+		} else if subnetDetails == nil {
+			// If the subnet was not found with provided ID, that is an error and a new subnet will not be created.
+			return false, fmt.Errorf("error failed to find subnet with id: %s", *subnet.ID)
+		}
+		return s.updateSubnetStatus(subnetDetails, isControlPlane)
+	} else if subnet.Name != nil {
+		// Attempt to check if a subnet exists with the name and update status as necessary.
+		subnetDetails, err := s.VPCClient.GetVPCSubnetByName(*subnet.Name)
+		if err != nil {
+			return false, fmt.Errorf("error retrieving subnet by name %s: %w", *subnet.Name, err)
+		} else if subnetDetails != nil {
+			// Update status if subnet was found.
+			return s.updateSubnetStatus(subnetDetails, isControlPlane)
+		}
+		// If subnet was not found, expect that it needs to be created.
+	} else {
+		// If no ID or name was provided, that is an error to be raised. One or the other must be specified when subnets are supplied.
+		return false, fmt.Errorf("error subnet has no defined id or name, one is required")
+	}
```

Comment:
this can be checked in the beginning and return early failure

Reply:
Okay, I think that is a safe expectation. updated